src/components/tl/nccl/tl_nccl_context.c

```diff
@@ -5,11 +5,71 @@
  */
 
 #include "tl_nccl.h"
+#include "core/ucc_mc.h"
+#include "core/ucc_ee.h"
+
+ucc_status_t ucc_tl_nccl_collective_progress(ucc_coll_task_t *coll_task)
+{
+    ucc_tl_nccl_task_t *task = ucc_derived_of(coll_task, ucc_tl_nccl_task_t);
+    ucc_status_t status;
+
+    status = ucc_mc_ee_event_test(task->completed, UCC_EE_CUDA_STREAM);
+    coll_task->super.status = status;
+    return status;
+}
+
+static void ucc_tl_nccl_req_mpool_obj_init(ucc_mpool_t *mp, void *obj,
+                                           void *chunk)
+{
+    ucc_tl_nccl_task_t *req = (ucc_tl_nccl_task_t*) obj;
+    req->super.progress = ucc_tl_nccl_collective_progress;
+}
+
 
 static ucc_mpool_ops_t ucc_tl_nccl_req_mpool_ops = {
     .chunk_alloc   = ucc_mpool_hugetlb_malloc,
     .chunk_release = ucc_mpool_hugetlb_free,
-    .obj_init      = NULL,
+    .obj_init      = ucc_tl_nccl_req_mpool_obj_init,
+    .obj_cleanup   = NULL
+};
+
+static ucc_status_t ucc_tl_nccl_req_mapped_mpool_chunk_malloc(ucc_mpool_t *mp,
+                                                              size_t *size_p,
+                                                              void ** chunk_p)
+{
+    cudaError_t cu_st;
+
+    cu_st = cudaHostAlloc((void**)chunk_p, *size_p, cudaHostAllocMapped);
+    if (cu_st != cudaSuccess) {
+        return UCC_ERR_NO_MEMORY;
+    }
+
+    return UCC_OK;
+}
+
+static void ucc_tl_nccl_req_mapped_mpool_chunk_free(ucc_mpool_t *mp,
+                                                    void *chunk)
+{
+    cudaFreeHost(chunk);
+}
+
+static void ucc_tl_nccl_req_mapped_mpool_obj_init(ucc_mpool_t *mp, void *obj,
+                                                  void *chunk)
+{
+    ucc_tl_nccl_task_t *req = (ucc_tl_nccl_task_t*) obj;
+    cudaError_t st;
+    st = cudaHostGetDevicePointer((void**)(&req->dev_status),
+                                  (void *)&req->super.super.status, 0);
+    if (st != cudaSuccess) {
+        req->super.super.status = UCC_ERR_NO_MESSAGE;
+    }
+    req->super.progress = NULL;
```

Comment:
progress is NULL, So, when does finalize() called in case of mem_ops?

Reply:
yes, progress is intentionally NULL, custreamwritevalue will update status and after that user might call finalize